modules/imgproc/src/pyramids.cpp

```diff
@@ -146,212 +133,146 @@ struct PyrDownVec_32f
 {
     int operator()(float** src, float* dst, int, int width) const
     {
-        if( !checkHardwareSupport(CV_CPU_SSE) )
-            return 0;
-
         int x = 0;
         const float *row0 = src[0], *row1 = src[1], *row2 = src[2], *row3 = src[3], *row4 = src[4];
-        __m128 _4 = _mm_set1_ps(4.f), _scale = _mm_set1_ps(1.f/256);
-        for( ; x <= width - 8; x += 8 )
+
+        v_float32 _4 = vx_setall_f32(4.f), _scale = vx_setall_f32(1.f/256);
+        for( ; x <= width - v_float32::nlanes; x += v_float32::nlanes)
         {
-            __m128 r0, r1, r2, r3, r4, t0, t1;
-            r0 = _mm_load_ps(row0 + x);
-            r1 = _mm_load_ps(row1 + x);
-            r2 = _mm_load_ps(row2 + x);
-            r3 = _mm_load_ps(row3 + x);
-            r4 = _mm_load_ps(row4 + x);
-            r0 = _mm_add_ps(r0, r4);
-            r1 = _mm_add_ps(_mm_add_ps(r1, r3), r2);
-            r0 = _mm_add_ps(r0, _mm_add_ps(r2, r2));
-            t0 = _mm_add_ps(r0, _mm_mul_ps(r1, _4));
-
-            r0 = _mm_load_ps(row0 + x + 4);
-            r1 = _mm_load_ps(row1 + x + 4);
-            r2 = _mm_load_ps(row2 + x + 4);
-            r3 = _mm_load_ps(row3 + x + 4);
-            r4 = _mm_load_ps(row4 + x + 4);
-            r0 = _mm_add_ps(r0, r4);
-            r1 = _mm_add_ps(_mm_add_ps(r1, r3), r2);
-            r0 = _mm_add_ps(r0, _mm_add_ps(r2, r2));
-            t1 = _mm_add_ps(r0, _mm_mul_ps(r1, _4));
-
-            t0 = _mm_mul_ps(t0, _scale);
-            t1 = _mm_mul_ps(t1, _scale);
-
-            _mm_storeu_ps(dst + x, t0);
-            _mm_storeu_ps(dst + x + 4, t1);
+            v_float32 r0, r1, r2, r3, r4;
+            r0 = vx_load(row0 + x);
+            r1 = vx_load(row1 + x);
+            r2 = vx_load(row2 + x);
+            r3 = vx_load(row3 + x);
+            r4 = vx_load(row4 + x);
+            v_store(dst + x, v_muladd(r1 + r3 + r2, _4, r0 + r4 + (r2 + r2)) * _scale);
         }
 
         return x;
     }
 };
 
-#if CV_SSE4_1
+#if CV_SSE4_1 || CV_NEON
```

Comment:
/cc @seiko2plus Do we need VSX here too?

Reply:
Yes, we need it